File: src/main/java/org/pixeltime/enchantmentsenhance/event/enchantment/Shooter.java
package org.pixeltime.enchantmentsenhance.event.enchantment;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.block.Block;
import org.bukkit.entity.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.entity.ProjectileHitEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.util.BlockIterator;
import org.pixeltime.enchantmentsenhance.Main;
import org.pixeltime.enchantmentsenhance.manager.SettingsManager;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.UUID;

public class Shooter implements Listener {
    private HashMap<Projectile, String> temp;
    private ArrayList<UUID> cooldown;

    public Shooter() {
        this.temp = new HashMap<Projectile, String>();
        this.cooldown = new ArrayList<UUID>();
    }

    @EventHandler
    public void onInteract(final PlayerInteractEvent playerInteractEvent) {
        if (playerInteractEvent.getAction() != Action.RIGHT_CLICK_AIR && playerInteractEvent.getAction() != Action.RIGHT_CLICK_BLOCK)
            return;
        final String translateAlternateColorCodes = ChatColor.translateAlternateColorCodes('&', SettingsManager.lang.getString("enchantment." + "shooter"));
        final Player player = playerInteractEvent.getPlayer();
        if (this.cooldown.contains(player.getUniqueId())) {
            return;
        }
        if (player.getItemInHand() != null && player.getItemInHand().hasItemMeta() && player.getItemInHand().getItemMeta().hasLore() && player.getItemInHand().getItemMeta().getLore().contains(String.valueOf(translateAlternateColorCodes) + " I")) {
            if (SettingsManager.enchant.getString("shooter.level_I.projectile").equalsIgnoreCase("snowball")) {
                final Snowball throwSnowball = player.throwSnowball();
                throwSnowball.setVelocity(player.getEyeLocation().getDirection().multiply(4));
                this.temp.put(throwSnowball, player.getName());
                if (SettingsManager.enchant.getBoolean("shooter.level_I.enable-cooldown")) {
                    final int n = SettingsManager.enchant.getInt("shooter.level_I.cooldown") * 20;
                    this.cooldown.add(player.getUniqueId());
                    new BukkitRunnable() {
                        public void run() {
                            Shooter.this.cooldown.remove(player.getUniqueId());
                        }
                    }.runTaskLater(Main.getMain(), (long) n);
                }
            } else if (SettingsManager.enchant.getString("shooter.level_I.projectile").equalsIgnoreCase("fireball")) {
                final Fireball fireball = (Fireball) player.launchProjectile((Class) Fireball.class);
                fireball.setIsIncendiary(false);
                fireball.setYield(0.0f);
                fireball.setVelocity(player.getEyeLocation().getDirection().multiply(4));
                this.temp.put(fireball, player.getName());
                if (SettingsManager.enchant.getBoolean("shooter.level_I.enable-cooldown")) {
                    final int n2 = SettingsManager.enchant.getInt("shooter.level_I.cooldown") * 20;
                    this.cooldown.add(player.getUniqueId());
                    new BukkitRunnable() {
                        public void run() {
                            Shooter.this.cooldown.remove(player.getUniqueId());
                        }
                    }.runTaskLater(Main.getMain(), (long) n2);
                }
            } else {
                if (!SettingsManager.enchant.getString("shooter.level_I.projectile").equalsIgnoreCase("egg")) {
                    return;
                }
                final Egg egg = (Egg) player.launchProjectile((Class) Egg.class);
                egg.setVelocity(player.getEyeLocation().getDirection().multiply(4));
                this.temp.put(egg, player.getName());
                if (SettingsManager.enchant.getBoolean("shooter.level_I.enable-cooldown")) {
                    final int n3 = SettingsManager.enchant.getInt("shooter.level_I.cooldown") * 20;
                    this.cooldown.add(player.getUniqueId());
                    new BukkitRunnable() {
                        public void run() {
                            Shooter.this.cooldown.remove(player.getUniqueId());
                        }
                    }.runTaskLater(Main.getMain(), (long) n3);
                }
            }
        }
    }

    @EventHandler
    public void onHit(final ProjectileHitEvent projectileHitEvent) {
        final Projectile entity = projectileHitEvent.getEntity();
        if (this.temp.containsKey(entity)) {
            final Player player = Bukkit.getPlayer(this.temp.get(entity));
            if (player == null) {
                return;
            }
            final BlockIterator blockIterator = new BlockIterator(projectileHitEvent.getEntity().getWorld(), projectileHitEvent.getEntity().getLocation().toVector(), projectileHitEvent.getEntity().getVelocity().normalize(), 0.0, 4);
            Block next = null;
            while (blockIterator.hasNext()) {
                next = blockIterator.next();
                if (next.getTypeId() != 0) {
                    break;
                }
            }
            final BlockBreakEvent blockBreakEvent = new BlockBreakEvent(next, player);
            Bukkit.getPluginManager().callEvent(blockBreakEvent);
            if (!blockBreakEvent.isCancelled()) {
                next.breakNaturally();
                this.temp.remove(entity);
            }
        }
    }
}

File: src/main/kotlin/org/pixeltime/enchantmentsenhance/event/enchantments/Shooter.kt
package org.pixeltime.enchantmentsenhance.event.enchantment

import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.block.Block
import org.bukkit.entity.*
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.entity.ProjectileHitEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.scheduler.BukkitRunnable
import org.bukkit.util.BlockIterator
import org.pixeltime.enchantmentsenhance.Main
import org.pixeltime.enchantmentsenhance.manager.SettingsManager

import java.util.ArrayList
import java.util.HashMap
import java.util.UUID

class Shooter : Listener {
    private val temp: HashMap<Projectile, String>
    private val cooldown: ArrayList<UUID>

    init {
        this.temp = HashMap()
        this.cooldown = ArrayList()
    }

    @EventHandler
    fun onInteract(playerInteractEvent: PlayerInteractEvent) {
        if (playerInteractEvent.action != Action.RIGHT_CLICK_AIR && playerInteractEvent.action != Action.RIGHT_CLICK_BLOCK)
            return
        val translateAlternateColorCodes = ChatColor.translateAlternateColorCodes('&', SettingsManager.lang.getString("enchantment." + "shooter"))
        val player = playerInteractEvent.player
        if (this.cooldown.contains(player.uniqueId)) {
            return
        }
        if (player.itemInHand != null && player.itemInHand.hasItemMeta() && player.itemInHand.itemMeta.hasLore() && player.itemInHand.itemMeta.lore.contains(translateAlternateColorCodes.toString() + " I")) {
            if (SettingsManager.enchant.getString("shooter.level_I.projectile").equals("snowball", ignoreCase = true)) {
                val throwSnowball = player.throwSnowball()
                throwSnowball.velocity = player.eyeLocation.direction.multiply(4)
                this.temp[throwSnowball] = player.name
                if (SettingsManager.enchant.getBoolean("shooter.level_I.enable-cooldown")) {
                    val n = SettingsManager.enchant.getInt("shooter.level_I.cooldown") * 20
                    this.cooldown.add(player.uniqueId)
                    object : BukkitRunnable() {
                        override fun run() {
                            this@Shooter.cooldown.remove(player.uniqueId)
                        }
                    }.runTaskLater(Main.getMain(), n.toLong())
                }
            } else if (SettingsManager.enchant.getString("shooter.level_I.projectile").equals("fireball", ignoreCase = true)) {
                val fireball = player.launchProjectile<Projectile>(Fireball::class.java as Class<*>) as Fireball
                fireball.setIsIncendiary(false)
                fireball.yield = 0.0f
                fireball.velocity = player.eyeLocation.direction.multiply(4)
                this.temp[fireball] = player.name
                if (SettingsManager.enchant.getBoolean("shooter.level_I.enable-cooldown")) {
                    val n2 = SettingsManager.enchant.getInt("shooter.level_I.cooldown") * 20
                    this.cooldown.add(player.uniqueId)
                    object : BukkitRunnable() {
                        override fun run() {
                            this@Shooter.cooldown.remove(player.uniqueId)
                        }
                    }.runTaskLater(Main.getMain(), n2.toLong())
                }
            } else {
                if (!SettingsManager.enchant.getString("shooter.level_I.projectile").equals("egg", ignoreCase = true)) {
                    return
                }
                val egg = player.launchProjectile<Projectile>(Egg::class.java as Class<*>) as Egg
                egg.velocity = player.eyeLocation.direction.multiply(4)
                this.temp[egg] = player.name
                if (SettingsManager.enchant.getBoolean("shooter.level_I.enable-cooldown")) {
                    val n3 = SettingsManager.enchant.getInt("shooter.level_I.cooldown") * 20
                    this.cooldown.add(player.uniqueId)
                    object : BukkitRunnable() {
                        override fun run() {
                            this@Shooter.cooldown.remove(player.uniqueId)
                        }
                    }.runTaskLater(Main.getMain(), n3.toLong())
                }
            }
        }
    }

    @EventHandler
    fun onHit(projectileHitEvent: ProjectileHitEvent) {
        val entity = projectileHitEvent.entity
        if (this.temp.containsKey(entity)) {
            val player = Bukkit.getPlayer(this.temp[entity]) ?: return
            val blockIterator = BlockIterator(projectileHitEvent.entity.world, projectileHitEvent.entity.location.toVector(), projectileHitEvent.entity.velocity.normalize(), 0.0, 4)
            var next: Block? = null
            while (blockIterator.hasNext()) {
                next = blockIterator.next()
                if (next!!.typeId != 0) {
                    break
                }
            }
            val blockBreakEvent = BlockBreakEvent(next, player)
            Bukkit.getPluginManager().callEvent(blockBreakEvent)
            if (!blockBreakEvent.isCancelled) {
                next!!.breakNaturally()
                this.temp.remove(entity)
            }
        }
    }
}
